src/device-api/actions.ts

```diff
@@ -36,3 +38,29 @@ export const identify = (ms: number = DEFAULT_IDENTIFY_DURATION) => {
 	blink.pattern.start();
 	setTimeout(blink.pattern.stop, ms);
 };
+
+/**
+ * Expires the supervisor's API key and generates a new one.
+ * Also communicates the new key to the balena API, if it's a key
+ * with global scope. The backend uses the global key to communicate
+ * with the Supervisor.
+ * Used by:
+ * - POST /v1/regenerate-api-key
+ */
+export const regenerateKey = async (oldKey: string) => {
+	try {
+		const shouldReportUpdatedKey = oldKey === (await getGlobalApiKey());
+		const newKey = await refreshKey(oldKey);
+
+		if (shouldReportUpdatedKey) {
+			deviceState.reportCurrentState({
+				api_secret: newKey,
+			});
+		}
+
+		return newKey;
+	} catch (e) {
+		log.error('Error regenerating API key', e);
```

Comment:
This interface is a bit weird, if an error occurred we  should just throw right? Why do we want to return the old key? This could be misleading, making the user of the method to think that the key changed when in fact it did not.

Reply:
I've changed the `regenerateKey` action to pass along the underlying `apiKey.refreshKey` error if one gets thrown, and the route `regenerate-api-key` to pass any errors to be handled by the error middleware.